### PR TITLE
TelemetryInitializer, Middleware, and TargetingContext rework

### PR DESCRIPTION
**Pull request by rossgrambo**
## Why this PR?

Primarily a refactor. 
Also resolves https://github.com/microsoft/FeatureManagement-Dotnet/issues/424. 
Related to- but explicitly not resolving here since the middleware was moved out of the main AspNetCore package https://github.com/microsoft/FeatureManagement-Dotnet/issues/396.

`ITargetingContextAccessor` is async and could not used reached by `ITelemetryInitializer`. We currently get around this by having our middleware add `TargetingContext` to `HttpContext` so it can be read synchronously.

This PR adjusts the middleware to instead place `TargetingId` on the `RequestTelemetry` object instead. This object is specific to Application Insights and we were already adding `TargetingId` to it with our existing initializer. By adding this earlier, we allow the initializer to depend on this field rather than the key on `HttpContext`. It also allows us to simply use the "TargetingId" key, rather than having a custom one for `HttpContext`.

This also allows `ITargetingContextAccessor` to cache however it wants and we're not caching in multiple places.

In short: Middleware stores on RequestTelemetry instead of HttpContext. Initializer now simply moves fields from RequestTelemetry to other Telemetry (like custom events).

## Visible Changes
`HttpContext` will no longer have the key "Microsoft.FeatureManagement.TargetingId" added to `HttpContext.Items`.

### CODE REVIEW

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/Constants.cs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore
+{
+    internal class Constants
+    {
+        public const string TargetingIdKey = "TargetingId";
```
Comment (zhiyuanliang-ms) [inline src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/Constants.cs:9]: Is there any reason why we discard the prefix "Microsoft.FeatureManagement" we used before?

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs [hunk 2]
```diff
@@ -50,7 +47,12 @@ public async Task InvokeAsync(HttpContext context, ITargetingContextAccessor tar
 
             if (targetingContext != null)
             {
-                context.Items[TargetingIdKey] = targetingContext.UserId;
+                var requestTelemetry = context.Features.Get<RequestTelemetry>();
+
+                if (requestTelemetry != null)
+                {
+                    requestTelemetry.Properties.Add(Constants.TargetingIdKey, targetingContext.UserId);
+                }
```
Comment (zhiyuanliang-ms) [inline src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs:55]: Should we also log something when requestTelemetry is missing?

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs [hunk 3]
```diff
@@ -42,20 +40,13 @@ protected override void OnInitializeTelemetry(HttpContext httpContext, RequestTe
...
-
-            if (!string.IsNullOrEmpty(targetingId))
+            // Extract the targeting id from the request telemetry
+            if (requestTelemetry.Properties.ContainsKey(Constants.TargetingIdKey))
             {
                 // Telemetry.Properties is deprecated in favor of ISupportProperties
                 if (telemetry is ISupportProperties telemetryWithSupportProperties)
                 {
-                    telemetryWithSupportProperties.Properties["TargetingId"] = targetingId;
+                    telemetryWithSupportProperties.Properties[Constants.TargetingIdKey] = requestTelemetry.Properties[Constants.TargetingIdKey];
```
Comment (zhiyuanliang-ms) [inline src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:49]: I think we can use `Add` method here.

This is a code snippet from `AspNetCoreEnvironmentTelemetryInitializer` from `Microsoft.ApplicationInsights.AspNetCore.TelemetryInitializers`
``` C#
public void Initialize(ITelemetry telemetry)
{
    if (environment != null && telemetry is ISupportProperties supportProperties && !supportProperties.Properties.ContainsKey("AspNetCoreEnvironment"))
    {
        supportProperties.Properties.Add("AspNetCoreEnvironment", environment.EnvironmentName);
    }
}
```
Let's follow it.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs [hunk 4]
```diff
@@ -50,7 +47,12 @@ public async Task InvokeAsync(HttpContext context, ITargetingContextAccessor tar
 
             if (targetingContext != null)
             {
-                context.Items[TargetingIdKey] = targetingContext.UserId;
+                var requestTelemetry = context.Features.Get<RequestTelemetry>();
+
+                if (requestTelemetry != null)
+                {
+                    requestTelemetry.Properties.Add(Constants.TargetingIdKey, targetingContext.UserId);
```
Comment (zhiyuanliang-ms) [inline src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs:54]: Before calling `Add` method, we need to do `ContainsKey` check.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/Constants.cs (hunk 1 above)
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/Constants.cs:9]: Before, we were placing it on `HttpContext.Items`. Many different services look at this and store their own things there- which is why it needed the full "Microsoft.FeatureManagement" prefix. 

Now, we're placing it in `RequestTelemetry`. Not only is the scope of this specific to App Insights, but more importantly "TargetingId" was already being placed on it, just later in the code (so this PR isn't introducing a new key). 

It's a reasonable argument to say we should add the prefix to "TargetingId" on all of the outgoing telemetry, which we might do at some point. But that decision should be a separate issue, as it wouldn't just live on "TargetingId" but all of the fields we emit (like "Variant" and "VariantAssignmentReason")

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs (hunk 2 above)
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs:55]: Thats a good idea- will do

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs (hunk 3 above)
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:49]: I think we have the correct usage right now. `.Add` will throw an exception if the key already exists. For our middleware, we really don't expect the key to already exist / it would be weird if it did. For this initializer, it runs on **all** telemetry, meaning it even runs on the RequestTelemetry itself. This means the key will already exist during that call.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs (hunk 4 above)
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs:54]: I suppose the question to me is- if the targeting id already exists, what should we do? I think there's 3 options:
1. Throw
2. Do nothing
3. Override

Depending on what we want- I think the right functions are:
1. `.Add`
2. `.TryAdd` or `ContainsKey`
3. `[...] = ...`

I could see an argument for all of them. Although I think Override might actually make the most sense to ensure our TargetingId tagging is reliable and if people add the middleware twice it's not breaking.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs [hunk 5]
```diff
@@ -42,20 +40,13 @@ protected override void OnInitializeTelemetry(HttpContext httpContext, RequestTe
                 throw new ArgumentNullException(nameof(httpContext));
             }
```
Comment (zhenlan) [inline src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:41]: Looks like `httpContext` is not needed anymore.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs [hunk 6]
```diff
@@ -50,7 +47,12 @@ public async Task InvokeAsync(HttpContext context, ITargetingContextAccessor tar
 
             if (targetingContext != null)
             {
-                context.Items[TargetingIdKey] = targetingContext.UserId;
+                var requestTelemetry = context.Features.Get<RequestTelemetry>();
```
Comment (zhenlan) [inline src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs:50]: NIT: can we rename `context` to `httpContext`? There are so many "contexts". Just make it a little bit easier to read the code :)

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs (hunk 4 above)
Comment (zhenlan) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs:54]: I don't think the middleware should do anything if the TargetingId is already there. We shouldn't just blindly override it. With that, I wouldn't even call `targetingContextAccessor.GetContextAsync()` if the TargetingId already exists.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs [hunk 7]
```diff
@@ -42,20 +40,13 @@ protected override void OnInitializeTelemetry(HttpContext httpContext, RequestTe
...
-            string targetingId = null;
-
-            if (httpContext.Items.TryGetValue(TargetingIdKey, out object value))
-            {
-                targetingId = value?.ToString();
-            }
-
-            if (!string.IsNullOrEmpty(targetingId))
+            // Extract the targeting id from the request telemetry
+            if (requestTelemetry.Properties.ContainsKey(Constants.TargetingIdKey))
```
Comment (zhenlan) [inline src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:44]: Should we do null check for `requestTelemetry` too?

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/Constants.cs (hunk 1 above)
Comment (zhenlan) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/Constants.cs:9]: I suppose this is the code you refer to
```cs
telemetryWithSupportProperties.Properties[Constants.TargetingIdKey] = requestTelemetry.Properties[Constants.TargetingIdKey];
```

The key on the left doesn't need to be the same key on the right. I agree the key on the left is a bigger discussion, but the key on the right can be just anything. Now that we put the middleware and initializer in the same package, the key on the right can even be something like `static object TargetingIdCacheKey = new object();`.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs (hunk 3 above)
Comment (zhenlan) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:49]: We shouldn't override it if someone already set it.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs (hunk 6 above)
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs:50]: Good call, yes.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs (hunk 7 above)
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:44]: Adjusted the `httpContext` null check to be a check on `requestTelemetry` instead.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs (hunk 4 above)
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingHttpContextMiddleware.cs]: Went ahead and adjusted this to `[...]=...` for now.

File: src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs (hunk 3 above)
Comment (rossgrambo) [reply src/Microsoft.FeatureManagement.Telemetry.ApplicationInsights.AspNetCore/TargetingTelemetryInitializer.cs:49]: @zhenlan  If someone already set it, what do you prefer happens in these cases: 
1. Developer is setting TargetingId correctly (I think throw, skip, or override are all valid)
2. Developer is setting TargetingId incorrectly. ie: to a different value than TargetingContext.UserId (I think throw or override).
3. Developer is not setting TargetingId (We add/override)

To me, override is the safest option. If they don't like it they can stop using the initializer.

**Comment by rossgrambo** [comment]
Met with Zhenlan offline. My takeaway is:

Should we be setting TargetingId on RequestTelemetry? 

If the answer is no- then this PR is not the correct way to solve this problem and we should use a cache. 
If the answer is yes- this PR is a valid way to persist the TargetingId to custom events without needing a cache. 

Our current preview release is placing TargetingId on the RequestTelemetry. However Split is not reading RequestTelemetry today, so it's not currently helpful to experimentation to be on there. 

Currently I'm on the fence if we should add it. It's not helpful to split yet- but it could be helpful in App Insights today. If we go with "no" and pursue a caching solution instead, would adding TargetingId to RequestTelemetry 6 months from now make the cache obsolete? If so, our caching solution better not be a breaking change to take away.

**Comment by zhiyuanliang-ms** [comment]
> Should we be setting TargetingId on RequestTelemetry at all?

Comparing to the current way (putting it in httpcontext.items), I will vote for setting TargetingId on RequestTelemetry. This is the recommanded way in App Insights [docs[(https://learn.microsoft.com/en-us/azure/azure-monitor/app/asp-net-core?tabs=netcorenew#enrich-data-through-http).

**Comment by zhenlan** [comment]
I am bit on the fence regarding whether we should add TargetingId to RequestTelemetry because we don't have any out-of-box usage scenario for RequestTelemetry today, but I'm not opposed to it. It could be handy for customers if they want to correlate feature evaluation with request telemetry. Of course, it could be handy for customers if we add TargetingId to all other ASP.NET telemetry (e.g. performance counters). I almost feel we should give customers an option to opt-in/out.

Regardless of whether we add TargetingId to RequestTelemetry (and other ASP.NET telemetry), it doesn't feel right to use RequestTelemetry as a caching mechanism in the middleware. We already have an initializer, and we must add TargetingId to the custom events. Even if we want to add TargetingId to RequestTelemetry (and any other ASP.NET telemetry), we should add it in the initializer.

**Comment by rossgrambo** [comment]
I agree, customers should probably be able to opt out of it. And if that's something we might offer- we would still need the custom events to be populated with it. I think that's a good argument for decoupling these. Going to close this PR.

**Comment by jimmyca15** [comment]
> I agree, customers should probably be able to opt out of it. And if that's something we might offer- we would still need the custom events to be populated with it. I think that's a good argument for decoupling these. Going to close this PR.

This PR moves the middleware into the asp.net core project though. Will you be opening a new PR for that?

**Comment by rossgrambo** [comment]
Yes I will. Going to solidify the shape of the caching first- since it has implications on what projects the middleware depends on. 

For example, with the changes in this PR- it made a lot of sense to place the middleware under App Insights. I want to make sure that's still the case after changes.
